service/dap/server.go

```diff
@@ -0,0 +1,464 @@
+// and is currently only called from the run goroutine.
+// TODO(polina): lock this when we add more goroutines that could call
+// this when we support asynchronous request-response communication.
+func (s *Server) signalDisconnect() {
+	// DisconnectChan might be nil at server creation if the
+	// caller does not want to rely on the disconnect signal.
+	if s.config.DisconnectChan != nil {
+		close(s.config.DisconnectChan)
+		// Take advantage of the nil check above to avoid accidentally
+		// closing the channel twice and causing a panic, when this
+		// function is called more than once. For example, we could
+		// have the following sequence of events:
+		// -- run goroutine: calls onDisconnectRequest()
+		// -- run goroutine: calls signalDisconnect()
+		// -- main goroutine: calls Stop()
+		// -- main goroutine: Stop() closes client connection
+		// -- run goroutine: serveDAPCodec() gets "closed network connection"
+		// -- run goroutine: serveDAPCodec() returns
+		// -- run goroutine: serveDAPCodec calls signalDisconnect()
+		s.config.DisconnectChan = nil
+	}
+}
+
+// Run launches a new goroutine where it accepts a client connection
+// and starts processing requests from it. Use Stop() to close connection.
+// The server does not support multiple clients, serially or in parallel.
+// The server should be restarted for every new debug session.
+// The debugger won't be started until launch/attach request is received.
+// TODO(polina): allow new client connections for new debug sessions,
+// so the editor needs to launch delve only once?
+func (s *Server) Run() {
+	go func() {
+		conn, err := s.listener.Accept()
+		if err != nil {
+			// This will print if the server is killed with Ctrl+C
+			// before client connection is accepted.
+			fmt.Printf("Error accepting client connection: %s\n", err)
+			s.signalDisconnect()
+			return
+		}
+		s.conn = conn
+		s.serveDAPCodec()
+	}()
+}
+
+// serveDAPCodec reads and decodes requests from the client
+// until it encounters an error or EOF, when it sends
+// the disconnect signal and returns.
+func (s *Server) serveDAPCodec() {
+	defer s.signalDisconnect()
+	s.reader = bufio.NewReader(s.conn)
+	for {
+		request, err := dap.ReadProtocolMessage(s.reader)
+		// TODO(polina): Differentiate between errors and handle them
+		// gracefully. For example,
+		// -- "use of closed network connection" means client connection
+		// was closed via Stop() in response to a disconnect request.
+		// -- "Request command 'foo' is not supported" means we
+		// potentially got some new DAP request that we do not yet have
+		// decoding support for, so we can respond with an ErrorResponse.
+		// TODO(polina): to support this add Seq to
+		// dap.DecodeProtocolMessageFieldError.
+		if err != nil {
+			if err != io.EOF {
+				fmt.Println("DAP error:", err)
+			}
+			return
+		}
+		s.handleRequest(request)
+	}
+}
+
+func (s *Server) handleRequest(request dap.Message) {
+	jsonmsg, _ := json.Marshal(request)
```

Comment:
I think that either this function or its caller should have some kind of panic guard, like service/rpccommon/server.go does (see newInternalError and its callers). The idea is that if we have a bug and we accidentally do an out of bounds slice access it's safer to recover from that panic and return it to the client (as an internal error) than to let the panic bring down Delve, because that will also kill the user's process, which is highly disruptive.

Reply:
I am going to add a TODO for now if that's OK. I have not yet implemented attach, so taking down the program as delve goes down comes with the limited experimental territory right now. I will need to figure out what returning the error to the client would mean here as the responses are always sent with the seq id of the request. And I can use the extra time to learn about this language feature (I am a go newbie).

Reply:
Ok.